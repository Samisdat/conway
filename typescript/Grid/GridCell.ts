import {Position} from '@Conway/Conway/Position';
import {CellTypesFactory} from '@Conway/CellType/CellTypesFactory';
import {CellType} from '@Conway/CellType/CellType';
import {CellUnkown} from '@Conway/CellType/CellUnkown';


export class GridCell {

    public readonly relativePosition: Position;
    public readonly absolutePosition: Position;
    public readonly offset: Position;

    private type: CellType = new CellUnkown();

    constructor(
        relativePosition: Position,
        absolutePosition: Position,
        offset: Position
    ) {

        this.relativePosition = relativePosition;

        this.absolutePosition = absolutePosition;

        this.offset = offset;

        this.setCheckerboardColor();

    }

    private setCheckerboardColor(): void {

        const x = this.absolutePosition.x;
        const y = this.absolutePosition.y;

        this.setType(CellTypesFactory.get().checkerboardDark());

        if (0 === y % 2) {
            if (0 === x % 2) {
                this.setType(CellTypesFactory.get().checkerboardLight());
            }
        }

        if (0 !== y % 2) {
            if (0 !== x % 2) {
                this.setType(CellTypesFactory.get().checkerboardLight());
            }
        }

        if (0 === x && 0 === y) {
            this.setType(CellTypesFactory.get().center());
        }

    }

    public setType(type: CellType): void {
        this.type = type;
    }

    public getType(): CellType {
        return this.type;
    }

    public getColor(): string {
        return this.type.hex;
    }

    get x(): number {
        return this.relativePosition.x + this.offset.x;
    }
    get y(): number {
        return this.relativePosition.y + this.offset.y;
    }

}
