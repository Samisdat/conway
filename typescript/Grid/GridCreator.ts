import { Position } from '../position';

export class GridCreator {


    public readonly canvasWidth:number;
    public readonly canvasHeigth:number;
    public readonly cellWidthAndHeight:number;
    public readonly pan:Position;

    private rows:number;
    private cols:number;

    constructor(
        canvasWidth:number,
        canvasHeigth:number,
        cellWidthAndHeight:number,
        pan:Position
    ) {
        this.canvasWidth = canvasWidth;
        this.canvasHeigth = canvasHeigth;
        this.cellWidthAndHeight = cellWidthAndHeight;
        this.pan = pan;


        this.rows = this.canvasWidth / cellWidthAndHeight;
        this.cols = this.canvasHeigth / cellWidthAndHeight;

    }

    public getRows():number{
        return this.rows;
    }

    public getCols():number{
        return this.cols;
    }

    public getPan():Position{
        return this.pan;
    }

}