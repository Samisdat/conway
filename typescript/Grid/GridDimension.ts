import {Integer} from '../Integer';

export class GridDimension {

    private readonly _rows: Integer;
    private readonly _cols: Integer;

    constructor(rows: number, cols: number) {

        this.ensurePositiveNumber(rows, 'rows');
        this.ensurePositiveNumber(cols, 'cols');

        this._rows = new Integer(rows);
        this._cols = new Integer(cols);

    }

    public get rows():number{
        return this._rows.value;
    }

    public get cols():number{
        return this._cols.value;
    }

    private ensurePositiveNumber(value: number, attributeName: string) {

        if (1 > value) {
            throw new Error(attributeName + ' must be at least 1');
        }


    }
}