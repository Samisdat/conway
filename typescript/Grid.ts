import { Position } from './Conway/position';
import { GridCell } from './Grid/GridCell';
import { Habitat } from './Conway/habitat';
import { CellTypeLiving } from 'CellType/CellTypeLiving';
import { CellTypesCenter } from 'CellType/CellTypesCenter';
import { GridDimension } from 'Grid/GridDimension';
import { IntegerPosition } from './Conway/IntegerPosition';
import {CellTypesFactory} from 'CellType/CellTypesFactory';

export class Grid {

    private readonly habitat: Habitat;

    private gridDimension: GridDimension;
    private sourcePosition: IntegerPosition;

    private offset: Position;

    private cells: GridCell[] = [];

    public readonly center: IntegerPosition;

    private absoluteMap: {} = {};

    constructor(
        habitat: Habitat,
        gridDimension: GridDimension,
        sourcePosition: IntegerPosition,
        offset: Position
    ) {

        this.ensureOffsetBetweenMinusOneAndOne(offset);

        this.habitat = habitat;

        this.gridDimension = gridDimension;

        this.sourcePosition = sourcePosition;

        this.offset = offset;

        // create rows

        let relativePointer = new IntegerPosition(0, 0);
        let absolutePointer = this.sourcePosition;

        this.center = relativePointer.clone().move(
            new IntegerPosition(
                Math.floor(this.gridDimension.rows / 2),
                Math.floor(this.gridDimension.cols / 2)
            )
        );

        relativePointer = relativePointer.move(
            new Position(
                -1 * Math.floor(this.gridDimension.rows / 2),
                -1 * Math.floor(this.gridDimension.cols / 2)
            )
        );

        absolutePointer = absolutePointer.move(
            new Position(
                -1 * Math.floor(this.gridDimension.rows / 2),
                -1 * Math.floor(this.gridDimension.cols / 2)
            )
        );

        const moveRight = new Position(
            1,
            0
        );

        // @TODO Use Bounds

        for (let top = 0; top < this.gridDimension.cols; top += 1) {

            for (let right = 0; right < this.gridDimension.rows; right += 1) {

                const gridCell = new GridCell(
                    relativePointer,
                    absolutePointer,
                    this.offset
                );

                const absolutePosition = gridCell.absolutePosition;

                if (undefined === this.absoluteMap[absolutePosition.x]) {
                    this.absoluteMap[absolutePosition.x] = {};
                }

                this.absoluteMap[absolutePosition.x][absolutePosition.y] = this.cells.length;

                this.cells.push(gridCell);

                relativePointer = relativePointer.move(
                    moveRight
                );

                absolutePointer = absolutePointer.move(moveRight);

            }

            relativePointer = relativePointer.move(
                new IntegerPosition(-1 * this.gridDimension.rows, 1),
            );

            absolutePointer = absolutePointer.move(
                new IntegerPosition(-1 * this.gridDimension.rows, 1),
            );

        }

        for (let positionWithLivingCells of habitat.get()) {

            if (undefined === this.absoluteMap[positionWithLivingCells.x]) {
                continue;
            }
            if (undefined === this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y]) {
                continue;
            }

            const cellIndex = this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y];

            this.cells[cellIndex].setType(CellTypesFactory.get().living());

        }

    }

    private ensureOffsetBetweenMinusOneAndOne(offset: Position) {

        if (1 < Math.abs(offset.x )) {
            throw new Error('offset.x must be between -1 and 1');
        }

        if (1 < Math.abs(offset.y)) {
            throw new Error('offset.y must be between -1 and 1');
        }


    }

    public getSourcePosition(): IntegerPosition {
        return this.sourcePosition;
    }

    public getOffset(): Position {
        return this.offset;
    }

    public getCols(): number {
        return this.gridDimension.cols;
    }

    public getRows(): number {
        return this.gridDimension.rows;
    }

    public getCells() {
        return this.cells;
    }

    public getCell(col: number, row: number): GridCell {

        for (const cell of this.cells) {

            if (cell.relativePosition.x === row && cell.relativePosition.y === col) {
                return cell;
            }
        }

        return undefined;

    }

    public getCellByAbsolutePosition(x: number, y: number): GridCell {

        if (undefined === this.absoluteMap[x]) {
            return undefined;
        }

        if (undefined === this.absoluteMap[x][y]) {
            return undefined;
        }

        const index = this.absoluteMap[x][y];

        return this.cells[index];
    }

}