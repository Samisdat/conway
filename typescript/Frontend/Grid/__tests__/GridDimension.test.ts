import {GridDimension} from '@Conway/Frontend/Grid/GridDimension';

describe('GridDimension', () => {

    test('can be created', () => {

        const gridDimension = new GridDimension(2, 3);

        expect(gridDimension).toBeInstanceOf(GridDimension);

    });


    test('rows can be retrieved', () => {

        const gridDimension = new GridDimension(2, 3);

        expect(gridDimension.rows).toBe(2);

    });

    test('cols can be retrieved', () => {

        const gridDimension = new GridDimension(2, 3);

        expect(gridDimension.cols).toBe(3);

    });

    test('throw execption for rows value less then 1', () => {

        expect(() => {
            new GridDimension(0, 3);
        }).toThrowErrorMatchingSnapshot();

        expect(() => {
            new GridDimension(-1, 3);
        }).toThrowErrorMatchingSnapshot();

    });

    test('throw exception for cols value less then 1', () => {

        expect(() => {
            new GridDimension(3, 0);
        }).toThrowErrorMatchingSnapshot();

        expect(() => {
            new GridDimension(3, -1);
        }).toThrowErrorMatchingSnapshot();

    });

});
