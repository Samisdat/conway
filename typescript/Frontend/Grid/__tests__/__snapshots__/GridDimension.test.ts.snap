// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridDimension throw exception for cols value less then 1 1`] = `"cols must be at least 1"`;

exports[`GridDimension throw exception for cols value less then 1 2`] = `"cols must be at least 1"`;

exports[`GridDimension throw execption for rows value less then 1 1`] = `"rows must be at least 1"`;

exports[`GridDimension throw execption for rows value less then 1 2`] = `"rows must be at least 1"`;
